File: smalivm/src/main/java/org/cf/smalivm/opcode/ReturnOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ReturnOp extends MethodStateOp {

    @SuppressWarnings("unused")
    private static final Logger log = LoggerFactory.getLogger(ReturnOp.class.getSimpleName());

    private final int register;

    ReturnOp(MethodLocation location, int register) {
        super(location);

        this.register = register;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        HeapItem item = mState.readRegister(register);
        mState.assignReturnRegister(item);
    }

    @Override
    public String toString() {
        return getName() + " r" + register;
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/ReturnOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.OpChild
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction11x
import org.slf4j.LoggerFactory

class ReturnOp internal constructor(location: MethodLocation, private val register: Int) : Op(location) {

    override val registersReadCount = 1
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out OpChild> {
        val value = node.state.readRegister(register)
        node.state.assignReturnRegister(value)
        return collectChildren()
    }

    override fun toString() = "$name r$register"

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(ReturnOp::class.java.simpleName)

        override fun build(
            location: MethodLocation,
            addressToLocation: Map<Int, MethodLocation>,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instr = location.instruction as Instruction11x
            val register = instr.registerA
            return ReturnOp(location, register)
        }
    }
}
